### ISSUE TITLE
Research SD.cpp

### ISSUE THREAD

**Issue by pminev** [alpaca-core/ac-local#191]
In the [Beta release](https://github.com/alpaca-core/ac-local/discussions/169) we're discussing to add Stable Diffusion support. Let's see what is the state of the library and are there any problems we might encounter.

**Comment by pminev** [comment]
I invest some time in Metal and for now it seems like there are some known/unknown issues. I've created an issue in their repo [read more here](https://github.com/leejet/stable-diffusion.cpp/issues/458). There I've explained what problems I've hit. To summarize it SD3 models generate black images with Metal while they work on CPU. I've tried it on CUDA and it was working properly as it was on the CPU, so it seems like there are some problems with Metal backend.

The only problems they have mentioned are mainly performance, but there are no notes for unsupported models or something like that. So I've decide to ask them to give us a little more information while I'm debugging a little bit to see where exactly are the problems.

**Comment by pminev** [comment]
For some reason the upsampling [calculation here](https://github.com/leejet/stable-diffusion.cpp/blob/ac54e0076052a196b7df961eb1f792c9ff4d7f22/vae.hpp#L414-L428) introduce the problems to `nan` results. 

Before that the tensor values seem to be correct and the same as CPU ones.

I tried the approach to set them as an output and get them after compute but then I received a GPU Address Fault Error and the command buffer execution is aborted. Will try to come up with another idea.

**Comment by pminev** [comment]
I caught the issue. The problem is one of the `im2col` tensor operations which are applied during the `upsampling`.
The flow starts from [here](https://github.com/leejet/stable-diffusion.cpp/blob/ac54e0076052a196b7df961eb1f792c9ff4d7f22/vae.hpp#L426) -> [here](https://github.com/leejet/stable-diffusion.cpp/blob/ac54e0076052a196b7df961eb1f792c9ff4d7f22/ggml_extend.hpp#L593) -> [im2col op](https://github.com/ggerganov/ggml/blob/e178a213638f7cebd96a6ddac0151351ca5da066/src/ggml.c#L3888). The problem happens only in the last upscaling from 512 to 1024.

<details>
<summary>Log from im2col operation on the GPU with Metal backend </summary>

```
  [0, 0, 0, 0] = 0.000000
  [0, 0, 0, 1] = 0.000000
  [0, 0, 0, 2] = 0.000000
  [0, 0, 0, 2301] = 0.000000
  [0, 0, 0, 2302] = 1.185547
  [0, 0, 0, 2303] = 1.185547
  [0, 0, 1, 0] = 0.000000
  [0, 0, 1, 1] = 0.000000
  [0, 0, 1, 2] = 0.000000
  [0, 0, 1, 2301] = 1.185547
  [0, 0, 1, 2302] = 1.185547
  [0, 0, 1, 2303] = 0.533691
  [0, 0, 2, 0] = 0.000000
  [0, 0, 2, 1] = 0.000000
  [0, 0, 2, 2] = 0.000000
  [0, 0, 2, 2301] = 1.185547
  [0, 0, 2, 2302] = 0.533691
  [0, 0, 2, 2303] = 0.533691
  [0, 0, 1021, 0] = 0.000000
  [0, 0, 1021, 1] = 0.000000
  [0, 0, 1021, 2] = 0.000000
  [0, 0, 1021, 2301] = 0.879883
  [0, 0, 1021, 2302] = 0.879883
  [0, 0, 1021, 2303] = 0.156616
  [0, 0, 1022, 0] = 0.000000
  [0, 0, 1022, 1] = 0.000000
  [0, 0, 1022, 2] = 0.000000
  [0, 0, 1022, 2301] = 0.879883
  [0, 0, 1022, 2302] = 0.156616
  [0, 0, 1022, 2303] = 0.156616
  [0, 0, 1023, 0] = 0.000000
  [0, 0, 1023, 1] = 0.000000
  [0, 0, 1023, 2] = 0.000000
  [0, 0, 1023, 2301] = 0.156616
  [0, 0, 1023, 2302] = 0.156616
  [0, 0, 1023, 2303] = 0.000000
  [0, 1, 0, 0] = 0.000000
  [0, 1, 0, 1] = -0.442627
  [0, 1, 0, 2] = -0.442627
  [0, 1, 0, 2301] = 0.000000
  [0, 1, 0, 2302] = 1.009766
  [0, 1, 0, 2303] = 1.009766
  [0, 1, 1, 0] = -0.442627
  [0, 1, 1, 1] = -0.442627
  [0, 1, 1, 2] = -0.185547
  [0, 1, 1, 2301] = 1.009766
  [0, 1, 1, 2302] = 1.009766
  [0, 1, 1, 2303] = 0.225342
  [0, 1, 2, 0] = -0.442627
  [0, 1, 2, 1] = -0.185547
  [0, 1, 2, 2] = -0.185547
  [0, 1, 2, 2301] = 1.009766
  [0, 1, 2, 2302] = 0.225342
  [0, 1, 2, 2303] = 0.225342
  [0, 1, 1021, 0] = 0.085083
  [0, 1, 1021, 1] = 0.085083
  [0, 1, 1021, 2] = 0.109253
  [0, 1, 1021, 2301] = 0.344482
  [0, 1, 1021, 2302] = 0.344482
  [0, 1, 1021, 2303] = -0.165894
  [0, 1, 1022, 0] = 0.085083
  [0, 1, 1022, 1] = 0.109253
  [0, 1, 1022, 2] = 0.109253
  [0, 1, 1022, 2301] = 0.344482
  [0, 1, 1022, 2302] = -0.165894
  [0, 1, 1022, 2303] = -0.165894
  [0, 1, 1023, 0] = 0.109253
  [0, 1, 1023, 1] = 0.109253
  [0, 1, 1023, 2] = 0.000000
  [0, 1, 1023, 2301] = -0.165894
  [0, 1, 1023, 2302] = -0.165894
  [0, 1, 1023, 2303] = 0.000000
  [0, 2, 0, 0] = 0.000000
  [0, 2, 0, 1] = -0.442627
  [0, 2, 0, 2] = -0.442627
  [0, 2, 0, 2301] = 0.000000
  [0, 2, 0, 2302] = 1.009766
  [0, 2, 0, 2303] = 1.009766
  [0, 2, 1, 0] = -0.442627
  [0, 2, 1, 1] = -0.442627
  [0, 2, 1, 2] = -0.185547
  [0, 2, 1, 2301] = 1.009766
  [0, 2, 1, 2302] = 1.009766
  [0, 2, 1, 2303] = 0.225342
  [0, 2, 2, 0] = -0.442627
  [0, 2, 2, 1] = -0.185547
  [0, 2, 2, 2] = -0.185547
  [0, 2, 2, 2301] = 1.009766
  [0, 2, 2, 2302] = 0.225342
  [0, 2, 2, 2303] = 0.225342
  [0, 2, 1021, 0] = 0.085083
  [0, 2, 1021, 1] = 0.085083
  [0, 2, 1021, 2] = 0.109253
  [0, 2, 1021, 2301] = 0.344482
  [0, 2, 1021, 2302] = 0.344482
  [0, 2, 1021, 2303] = -0.165894
  [0, 2, 1022, 0] = 0.085083
  [0, 2, 1022, 1] = 0.109253
  [0, 2, 1022, 2] = 0.109253
  [0, 2, 1022, 2301] = 0.344482
  [0, 2, 1022, 2302] = -0.165894
  [0, 2, 1022, 2303] = -0.165894
  [0, 2, 1023, 0] = 0.109253
  [0, 2, 1023, 1] = 0.109253
  [0, 2, 1023, 2] = 0.000000
  [0, 2, 1023, 2301] = -0.165894
  [0, 2, 1023, 2302] = -0.165894
  [0, 2, 1023, 2303] = 0.000000
  [0, 1021, 0, 0] = 0.000000
  [0, 1021, 0, 1] = 0.000000
  [0, 1021, 0, 2] = 0.000000
  [0, 1021, 0, 2301] = 0.000000
  [0, 1021, 0, 2302] = 0.000000
  [0, 1021, 0, 2303] = 0.000000
  [0, 1021, 1, 0] = 0.000000
  [0, 1021, 1, 1] = 0.000000
  [0, 1021, 1, 2] = 0.000000
  [0, 1021, 1, 2301] = 0.000000
  [0, 1021, 1, 2302] = 0.000000
  [0, 1021, 1, 2303] = 0.000000
  [0, 1021, 2, 0] = 0.000000
  [0, 1021, 2, 1] = 0.000000
  [0, 1021, 2, 2] = 0.000000
  [0, 1021, 2, 2301] = 0.000000
  [0, 1021, 2, 2302] = 0.000000
  [0, 1021, 2, 2303] = 0.000000
  [0, 1021, 1021, 0] = 0.000000
  [0, 1021, 1021, 1] = 0.000000
  [0, 1021, 1021, 2] = 0.000000
  [0, 1021, 1021, 2301] = 0.000000
  [0, 1021, 1021, 2302] = 0.000000
  [0, 1021, 1021, 2303] = 0.000000
  [0, 1021, 1022, 0] = 0.000000
  [0, 1021, 1022, 1] = 0.000000
  [0, 1021, 1022, 2] = 0.000000
  [0, 1021, 1022, 2301] = 0.000000
  [0, 1021, 1022, 2302] = 0.000000
  [0, 1021, 1022, 2303] = 0.000000
  [0, 1021, 1023, 0] = 0.000000
  [0, 1021, 1023, 1] = 0.000000
  [0, 1021, 1023, 2] = 0.000000
  [0, 1021, 1023, 2301] = 0.000000
  [0, 1021, 1023, 2302] = 0.000000
  [0, 1021, 1023, 2303] = 0.000000
  [0, 1022, 0, 0] = 0.000000
  [0, 1022, 0, 1] = 0.000000
  [0, 1022, 0, 2] = 0.000000
  [0, 1022, 0, 2301] = 0.000000
  [0, 1022, 0, 2302] = 0.000000
  [0, 1022, 0, 2303] = 0.000000
  [0, 1022, 1, 0] = 0.000000
  [0, 1022, 1, 1] = 0.000000
  [0, 1022, 1, 2] = 0.000000
  [0, 1022, 1, 2301] = 0.000000
  [0, 1022, 1, 2302] = 0.000000
  [0, 1022, 1, 2303] = 0.000000
  [0, 1022, 2, 0] = 0.000000
  [0, 1022, 2, 1] = 0.000000
  [0, 1022, 2, 2] = 0.000000
  [0, 1022, 2, 2301] = 0.000000
  [0, 1022, 2, 2302] = 0.000000
  [0, 1022, 2, 2303] = 0.000000
  [0, 1022, 1021, 0] = 0.000000
  [0, 1022, 1021, 1] = 0.000000
  [0, 1022, 1021, 2] = 0.000000
  [0, 1022, 1021, 2301] = 0.000000
  [0, 1022, 1021, 2302] = 0.000000
  [0, 1022, 1021, 2303] = 0.000000
  [0, 1022, 1022, 0] = 0.000000
  [0, 1022, 1022, 1] = 0.000000
  [0, 1022, 1022, 2] = 0.000000
  [0, 1022, 1022, 2301] = 0.000000
  [0, 1022, 1022, 2302] = 0.000000
  [0, 1022, 1022, 2303] = 0.000000
  [0, 1022, 1023, 0] = 0.000000
  [0, 1022, 1023, 1] = 0.000000
  [0, 1022, 1023, 2] = 0.000000
  [0, 1022, 1023, 2301] = 0.000000
  [0, 1022, 1023, 2302] = 0.000000
  [0, 1022, 1023, 2303] = 0.000000
  [0, 1023, 0, 0] = 0.000000
  [0, 1023, 0, 1] = 0.000000
  [0, 1023, 0, 2] = 0.000000
  [0, 1023, 0, 2301] = 0.000000
  [0, 1023, 0, 2302] = 0.000000
  [0, 1023, 0, 2303] = 0.000000
  [0, 1023, 1, 0] = 0.000000
  [0, 1023, 1, 1] = 0.000000
  [0, 1023, 1, 2] = 0.000000
  [0, 1023, 1, 2301] = 0.000000
  [0, 1023, 1, 2302] = 0.000000
  [0, 1023, 1, 2303] = 0.000000
  [0, 1023, 2, 0] = 0.000000
  [0, 1023, 2, 1] = 0.000000
  [0, 1023, 2, 2] = 0.000000
  [0, 1023, 2, 2301] = 0.000000
  [0, 1023, 2, 2302] = 0.000000
  [0, 1023, 2, 2303] = 0.000000
  [0, 1023, 1021, 0] = 0.000000
  [0, 1023, 1021, 1] = 0.000000
  [0, 1023, 1021, 2] = 0.000000
  [0, 1023, 1021, 2301] = 0.000000
  [0, 1023, 1021, 2302] = 0.000000
  [0, 1023, 1021, 2303] = 0.000000
  [0, 1023, 1022, 0] = 0.000000
  [0, 1023, 1022, 1] = 0.000000
  [0, 1023, 1022, 2] = 0.000000
  [0, 1023, 1022, 2301] = 0.000000
  [0, 1023, 1022, 2302] = 0.000000
  [0, 1023, 1022, 2303] = 0.000000
  [0, 1023, 1023, 0] = 0.000000
  [0, 1023, 1023, 1] = 0.000000
  [0, 1023, 1023, 2] = 0.000000
  [0, 1023, 1023, 2301] = 0.000000
  [0, 1023, 1023, 2302] = 0.000000
  [0, 1023, 1023, 2303] = 0.000000

```

</details>

<details>
<summary>Log from im2col operation on the CPU </summary>

```
  [0, 0, 0, 0] = 0.000000
  [0, 0, 0, 1] = 0.000000
  [0, 0, 0, 2] = 0.000000
  [0, 0, 0, 2301] = 0.000000
  [0, 0, 0, 2302] = 1.178711
  [0, 0, 0, 2303] = 1.178711
  [0, 0, 1, 0] = 0.000000
  [0, 0, 1, 1] = 0.000000
  [0, 0, 1, 2] = 0.000000
  [0, 0, 1, 2301] = 1.178711
  [0, 0, 1, 2302] = 1.178711
  [0, 0, 1, 2303] = 0.530273
  [0, 0, 2, 0] = 0.000000
  [0, 0, 2, 1] = 0.000000
  [0, 0, 2, 2] = 0.000000
  [0, 0, 2, 2301] = 1.178711
  [0, 0, 2, 2302] = 0.530273
  [0, 0, 2, 2303] = 0.530273
  [0, 0, 1021, 0] = 0.000000
  [0, 0, 1021, 1] = 0.000000
  [0, 0, 1021, 2] = 0.000000
  [0, 0, 1021, 2301] = 0.883789
  [0, 0, 1021, 2302] = 0.883789
  [0, 0, 1021, 2303] = 0.156982
  [0, 0, 1022, 0] = 0.000000
  [0, 0, 1022, 1] = 0.000000
  [0, 0, 1022, 2] = 0.000000
  [0, 0, 1022, 2301] = 0.883789
  [0, 0, 1022, 2302] = 0.156982
  [0, 0, 1022, 2303] = 0.156982
  [0, 0, 1023, 0] = 0.000000
  [0, 0, 1023, 1] = 0.000000
  [0, 0, 1023, 2] = 0.000000
  [0, 0, 1023, 2301] = 0.156982
  [0, 0, 1023, 2302] = 0.156982
  [0, 0, 1023, 2303] = 0.000000
  [0, 1, 0, 0] = 0.000000
  [0, 1, 0, 1] = -0.449463
  [0, 1, 0, 2] = -0.449463
  [0, 1, 0, 2301] = 0.000000
  [0, 1, 0, 2302] = 0.998047
  [0, 1, 0, 2303] = 0.998047
  [0, 1, 1, 0] = -0.449463
  [0, 1, 1, 1] = -0.449463
  [0, 1, 1, 2] = -0.182007
  [0, 1, 1, 2301] = 0.998047
  [0, 1, 1, 2302] = 0.998047
  [0, 1, 1, 2303] = 0.219971
  [0, 1, 2, 0] = -0.449463
  [0, 1, 2, 1] = -0.182007
  [0, 1, 2, 2] = -0.182007
  [0, 1, 2, 2301] = 0.998047
  [0, 1, 2, 2302] = 0.219971
  [0, 1, 2, 2303] = 0.219971
  [0, 1, 1021, 0] = 0.083984
  [0, 1, 1021, 1] = 0.083984
  [0, 1, 1021, 2] = 0.102417
  [0, 1, 1021, 2301] = 0.347168
  [0, 1, 1021, 2302] = 0.347168
  [0, 1, 1021, 2303] = -0.165894
  [0, 1, 1022, 0] = 0.083984
  [0, 1, 1022, 1] = 0.102417
  [0, 1, 1022, 2] = 0.102417
  [0, 1, 1022, 2301] = 0.347168
  [0, 1, 1022, 2302] = -0.165894
  [0, 1, 1022, 2303] = -0.165894
  [0, 1, 1023, 0] = 0.102417
  [0, 1, 1023, 1] = 0.102417
  [0, 1, 1023, 2] = 0.000000
  [0, 1, 1023, 2301] = -0.165894
  [0, 1, 1023, 2302] = -0.165894
  [0, 1, 1023, 2303] = 0.000000
  [0, 2, 0, 0] = 0.000000
  [0, 2, 0, 1] = -0.449463
  [0, 2, 0, 2] = -0.449463
  [0, 2, 0, 2301] = 0.000000
  [0, 2, 0, 2302] = 0.998047
  [0, 2, 0, 2303] = 0.998047
  [0, 2, 1, 0] = -0.449463
  [0, 2, 1, 1] = -0.449463
  [0, 2, 1, 2] = -0.182007
  [0, 2, 1, 2301] = 0.998047
  [0, 2, 1, 2302] = 0.998047
  [0, 2, 1, 2303] = 0.219971
  [0, 2, 2, 0] = -0.449463
  [0, 2, 2, 1] = -0.182007
  [0, 2, 2, 2] = -0.182007
  [0, 2, 2, 2301] = 0.998047
  [0, 2, 2, 2302] = 0.219971
  [0, 2, 2, 2303] = 0.219971
  [0, 2, 1021, 0] = 0.083984
  [0, 2, 1021, 1] = 0.083984
  [0, 2, 1021, 2] = 0.102417
  [0, 2, 1021, 2301] = 0.347168
  [0, 2, 1021, 2302] = 0.347168
  [0, 2, 1021, 2303] = -0.165894
  [0, 2, 1022, 0] = 0.083984
  [0, 2, 1022, 1] = 0.102417
  [0, 2, 1022, 2] = 0.102417
  [0, 2, 1022, 2301] = 0.347168
  [0, 2, 1022, 2302] = -0.165894
  [0, 2, 1022, 2303] = -0.165894
  [0, 2, 1023, 0] = 0.102417
  [0, 2, 1023, 1] = 0.102417
  [0, 2, 1023, 2] = 0.000000
  [0, 2, 1023, 2301] = -0.165894
  [0, 2, 1023, 2302] = -0.165894
  [0, 2, 1023, 2303] = 0.000000
  [0, 1021, 0, 0] = 0.000000
  [0, 1021, 0, 1] = 0.126343
  [0, 1021, 0, 2] = 0.126343
  [0, 1021, 0, 2301] = 0.000000
  [0, 1021, 0, 2302] = 0.389893
  [0, 1021, 0, 2303] = 0.389893
  [0, 1021, 1, 0] = 0.126343
  [0, 1021, 1, 1] = 0.126343
  [0, 1021, 1, 2] = 0.573242
  [0, 1021, 1, 2301] = 0.389893
  [0, 1021, 1, 2302] = 0.389893
  [0, 1021, 1, 2303] = 0.156982
  [0, 1021, 2, 0] = 0.126343
  [0, 1021, 2, 1] = 0.573242
  [0, 1021, 2, 2] = 0.573242
  [0, 1021, 2, 2301] = 0.389893
  [0, 1021, 2, 2302] = 0.156982
  [0, 1021, 2, 2303] = 0.156982
  [0, 1021, 1021, 0] = 0.298828
  [0, 1021, 1021, 1] = 0.298828
  [0, 1021, 1021, 2] = 0.416748
  [0, 1021, 1021, 2301] = 0.065308
  [0, 1021, 1021, 2302] = 0.065308
  [0, 1021, 1021, 2303] = 0.035919
  [0, 1021, 1022, 0] = 0.298828
  [0, 1021, 1022, 1] = 0.416748
  [0, 1021, 1022, 2] = 0.416748
  [0, 1021, 1022, 2301] = 0.065308
  [0, 1021, 1022, 2302] = 0.035919
  [0, 1021, 1022, 2303] = 0.035919
  [0, 1021, 1023, 0] = 0.416748
  [0, 1021, 1023, 1] = 0.416748
  [0, 1021, 1023, 2] = 0.000000
  [0, 1021, 1023, 2301] = 0.035919
  [0, 1021, 1023, 2302] = 0.035919
  [0, 1021, 1023, 2303] = 0.000000
  [0, 1022, 0, 0] = 0.000000
  [0, 1022, 0, 1] = 0.126343
  [0, 1022, 0, 2] = 0.126343
  [0, 1022, 0, 2301] = 0.000000
  [0, 1022, 0, 2302] = 0.389893
  [0, 1022, 0, 2303] = 0.389893
  [0, 1022, 1, 0] = 0.126343
  [0, 1022, 1, 1] = 0.126343
  [0, 1022, 1, 2] = 0.573242
  [0, 1022, 1, 2301] = 0.389893
  [0, 1022, 1, 2302] = 0.389893
  [0, 1022, 1, 2303] = 0.156982
  [0, 1022, 2, 0] = 0.126343
  [0, 1022, 2, 1] = 0.573242
  [0, 1022, 2, 2] = 0.573242
  [0, 1022, 2, 2301] = 0.389893
  [0, 1022, 2, 2302] = 0.156982
  [0, 1022, 2, 2303] = 0.156982
  [0, 1022, 1021, 0] = 0.298828
  [0, 1022, 1021, 1] = 0.298828
  [0, 1022, 1021, 2] = 0.416748
  [0, 1022, 1021, 2301] = 0.065308
  [0, 1022, 1021, 2302] = 0.065308
  [0, 1022, 1021, 2303] = 0.035919
  [0, 1022, 1022, 0] = 0.298828
  [0, 1022, 1022, 1] = 0.416748
  [0, 1022, 1022, 2] = 0.416748
  [0, 1022, 1022, 2301] = 0.065308
  [0, 1022, 1022, 2302] = 0.035919
  [0, 1022, 1022, 2303] = 0.035919
  [0, 1022, 1023, 0] = 0.416748
  [0, 1022, 1023, 1] = 0.416748
  [0, 1022, 1023, 2] = 0.000000
  [0, 1022, 1023, 2301] = 0.035919
  [0, 1022, 1023, 2302] = 0.035919
  [0, 1022, 1023, 2303] = 0.000000
  [0, 1023, 0, 0] = 0.000000
  [0, 1023, 0, 1] = -0.039917
  [0, 1023, 0, 2] = -0.039917
  [0, 1023, 0, 2301] = 0.000000
  [0, 1023, 0, 2302] = 0.000000
  [0, 1023, 0, 2303] = 0.000000
  [0, 1023, 1, 0] = -0.039917
  [0, 1023, 1, 1] = -0.039917
  [0, 1023, 1, 2] = -0.231689
  [0, 1023, 1, 2301] = 0.000000
  [0, 1023, 1, 2302] = 0.000000
  [0, 1023, 1, 2303] = 0.000000
  [0, 1023, 2, 0] = -0.039917
  [0, 1023, 2, 1] = -0.231689
  [0, 1023, 2, 2] = -0.231689
  [0, 1023, 2, 2301] = 0.000000
  [0, 1023, 2, 2302] = 0.000000
  [0, 1023, 2, 2303] = 0.000000
  [0, 1023, 1021, 0] = -0.022537
  [0, 1023, 1021, 1] = -0.022537
  [0, 1023, 1021, 2] = -0.082764
  [0, 1023, 1021, 2301] = 0.000000
  [0, 1023, 1021, 2302] = 0.000000
  [0, 1023, 1021, 2303] = 0.000000
  [0, 1023, 1022, 0] = -0.022537
  [0, 1023, 1022, 1] = -0.082764
  [0, 1023, 1022, 2] = -0.082764
  [0, 1023, 1022, 2301] = 0.000000
  [0, 1023, 1022, 2302] = 0.000000
  [0, 1023, 1022, 2303] = 0.000000
  [0, 1023, 1023, 0] = -0.082764
  [0, 1023, 1023, 1] = -0.082764
  [0, 1023, 1023, 2] = 0.000000
  [0, 1023, 1023, 2301] = 0.000000
  [0, 1023, 1023, 2302] = 0.000000
  [0, 1023, 1023, 2303] = 0.000000

```

</details>

**Comment by pminev** [comment]
Changed the test for `im2col` operation. [Here](https://github.com/ggerganov/ggml/blob/e178a213638f7cebd96a6ddac0151351ca5da066/tests/test-backend-ops.cpp#L2319)


I set the following values which are from the stable diffusion  upscaling:

```
...
            std::array<int64_t, 4> ne_input = {1024, 1024, 256, 1}, // [input_width, input_height, input_channels, 1]
            std::array<int64_t, 4> ne_kernel = {3, 3, 256, 256}, // [kernel_width, kernel_height, input_channels, 1]
...
``` 

And the tests started to fail. So, it seems like the implementation in ggml is having a bug.

**Comment by pminev** [comment]
Finally, i was able to fix im2col operation after which there was another bug in matrix multiplication, but it was something little (an integer overflow), so it was easy to fix and as a result we can run stable diffusion on Metal.
Final result.
![output](https://github.com/user-attachments/assets/75953a7d-3db6-41ff-96f4-ec3d9a883c46)

I tested it on Cuda, it was working properly - will double check because I'm not sure if I've tested with the same model.
On the CPU it was working, but slow. For the cat from above results are:
- Metal - 342 sec in total
- CPU - 1217 sec in total
